core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+
+    private ThreadPool threadPool;
+
+    @Before
+    public void setup() {
+        threadPool = new ThreadPool(Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "fixed delay tests").build());
+    }
+
+    @After
+    public void shutdown() {
+        threadPool.shutdownNow();
+    }
+
+    public void testDoesNotRescheduleUntilExecutionFinished() throws Exception {
+        final TimeValue delay = TimeValue.timeValueMillis(100L);
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch pauseLatch = new CountDownLatch(1);
+        ThreadPool threadPool = mock(ThreadPool.class);
+        final Runnable runnable = () ->  {
+            // notify that the runnable is started
+            startLatch.countDown();
+            try {
+                // wait for other thread to un-pause
+                pauseLatch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        };
+        ReschedulingRunnable reschedulingRunnable = new ReschedulingRunnable(runnable, delay, Names.GENERIC, threadPool);
+        // this call was made during construction of the runnable
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+
+        // create a thread and start the runnable
+        Thread runThread = new Thread() {
+            @Override
+            public void run() {
+                reschedulingRunnable.run();
+            }
+        };
+        runThread.start();
+
+        // wait for the runnable to be started and ensure the runnable hasn't used the threadpool again
+        startLatch.await();
+        verifyNoMoreInteractions(threadPool);
+
+        // un-pause the runnable and allow it to complete execution
+        pauseLatch.countDown();
+        runThread.join();
+
+        // validate schedule was called again
+        verify(threadPool, times(2)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+    }
+
+    public void testThatRunnableIsRescheduled() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(scaledRandomIntBetween(2, 16));
+        final Runnable countingRunnable = () -> {
+            if (rarely()) {
+                throw new ElasticsearchException("sometimes we throw before counting down");
+            }
+
+            latch.countDown();
+
+            if (randomBoolean()) {
+                throw new ElasticsearchException("this shouldn't cause the test to fail!");
+            }
+        };
+
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(100L), Names.GENERIC);
+        assertNotNull(cancellable);
+
+        // wait for the number of successful count down operations
+        latch.await();
+
+        // cancel
+        cancellable.cancel();
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testCancellingRunnable() throws Exception {
```

Comment:
I'm fine we keeping the test of canceling from the runnable task, but can we also have a test for external canceling where post cancel call the runnable is a runned at most once? Note that to do this you will again be in that situation where you wait for something that shouldn't happen, causing the test to be slow - I'm fine with a direct check that will sometime cause false positives.